crates/rust-analyzer/src/cli/lsif.rs

```diff
@@ -0,0 +1,146 @@
+        Snap(self.0.snapshot())
+    }
+}
+
+mod lsif_types;
+use lsif_types::*;
+
+#[derive(Default)]
+struct LsifManager {
+    count: i32,
+}
+
+#[derive(Clone, Copy)]
+struct Id(i32);
+
+impl From<Id> for NumberOrString {
+    fn from(Id(x): Id) -> Self {
+        NumberOrString::Number(x)
+    }
+}
+
+impl LsifManager {
+    fn add(&mut self, data: Element) -> Id {
+        let id = Id(self.count);
+        self.emit(&serde_json::to_string(&Entry { id: id.into(), data }).unwrap());
+        self.count += 1;
+        id
+    }
+
+    // TODO: support file in addition to stdout here
+    fn emit(&self, data: &str) {
+        println!("{}", data);
+    }
+}
+
+impl flags::Lsif {
+    pub fn run(self) -> Result<()> {
+        let cargo_config = CargoConfig::default();
+        let no_progress = &|_| ();
+        let load_cargo_config = LoadCargoConfig {
+            load_out_dirs_from_check: true,
+            with_proc_macro: true,
+            prefill_caches: false,
+        };
+        let path = AbsPathBuf::assert(env::current_dir()?.join(&self.path));
+        let manifest = ProjectManifest::discover_single(&path)?;
+
+        let workspace = ProjectWorkspace::load(manifest, &cargo_config, no_progress)?;
+
+        let (host, vfs, _proc_macro) = load_workspace(workspace, &load_cargo_config)?;
+        let db = host.raw_database();
+        let analysis = host.analysis();
+
+        let work = all_modules(db).into_iter().filter(|module| {
+            let file_id = module.definition_source(db).file_id.original_file(db);
+            let source_root = db.file_source_root(file_id);
+            let source_root = db.source_root(source_root);
+            !source_root.is_library
+        });
+
+        let mut visited_files = FxHashSet::default();
+
+        let mut lsif = LsifManager::default();
+
+        lsif.add(Element::Vertex(Vertex::MetaData {
+            version: String::from("0.5.0"),
+            project_root: lsp_types::Url::from_file_path(path).unwrap(),
+            position_encoding: Encoding::Utf16,
+            tool_info: None,
+        }));
+
+        for module in work {
+            let file_id = module.definition_source(db).file_id.original_file(db);
+            if !visited_files.contains(&file_id) {
+                let path = vfs.file_path(file_id);
+                let path = path.as_path().unwrap();
+                let doc_id = lsif.add(Element::Vertex(Vertex::Document(Document {
+                    language_id: Language::Rust,
+                    uri: lsp_types::Url::from_file_path(path).unwrap(),
+                })));
+                let folds = analysis.folding_ranges(file_id)?;
+                let text = analysis.file_text(file_id)?;
+                let line_index = db.line_index(file_id);
+                let result = folds
+                    .into_iter()
+                    .map(|it| to_proto::folding_range(&*text, &LineIndex::with_default_options(line_index.clone()), false, it))
```

Comment:
This very likely needs to use UTF-16 and not UTF-8. It also needs the actual line endings of this file.

Reply:
Thanks for your explanation, I didn't know what are those and I just put it there for silencing the compiler.

How I can get the actual line endings of the file?

Reply:
https://github.com/rust-analyzer/rust-analyzer/blob/2c538e9c8c622123853d94ac4827d6a41dded447/crates/rust-analyzer/src/global_state.rs#L304-L309 I think you can't call this function, but you can copy it's implementation.

Reply:
I don't have `vfs.read().1` without a global_state. Seems line endings are being calculated via `LineEndings::normalize`. This function also changes the text, I guess I should use the output of this file instead of original text, Am I right?

Reply:
`analysis.file_text()` gives the text after `LineEndings::normalize`, you need what kind of line ending is used. You do have `vfs` though. `load_workspace` returns it.

Reply:
`vfs` name here is misleading. Mine from `load_workspace` has type `Vfs` but the one in global_state has type `Arc<RwLock<RawRwLock, (Vfs, HashMap<FileId, LineEndings, BuildHasherDefault<FxHasher>>)>>` so `vfs.read().1` is `HashMap<FileId, LineEndings, BuildHasherDefault<FxHasher>` which I don't have. If `analysis.file_text()` is for after normalizing then my text is useless and I can't call normalize. So what to do?

Reply:
@matklad what is the best way to get line endings?

Reply:
Why do we need to know about line endings at all here? LSIF use line/column as a coordinate space, so it shouldn't care about those things.

Reply:
So I should set Unix as default? And what is difference between LSIF and LSP in this?

Reply:
LSP supports edits, and, to emit edits, we need to know line-endings. I think that's the only difference, but better to check usages of line-ending info.

Using unix should be fine, long-term, might make sense to refactor this somewhat to not give line-endings to stuff that doesn't need them.